### ISSUE TITLE
Method to automate the exposure of an entire class's methods

### ISSUE THREAD

**Issue by flatmax** [vphantom/js-jrpc#9]
Hi there.

Is it possible to "expose" all the methods of a given class ?

thanks
Matt

**Comment by vphantom** [comment]
Hi. :)

I assume that you have a use case where the `expose(object)` signature https://github.com/vphantom/js-jrpc#remoteexposeobject doesn't solve what you need?

**Comment by flatmax** [comment]
That method requires the object to conform to the jsrpc stringified 
method names + function prototype with input arguments params and next 
... so, for that reason, it doesn't really take an arbitrary object and 
expose it.

Would be very interested in that feature.

Matt

**Comment by vphantom** [comment]
Oh I see, you'd want a generic wrapper to handle the async aspect, and expose sync methods.  With binding and promises, that might be possible somewhat elegantly.  It'd stall the event loop if you're doing sync I/O, but if they're small methods I could see a use for that.

Thanks for the suggestion; I'll give that some thought.  In the mean time, feel free to add an example of what you have in mind to make sure I understand your intent correctly.

**Comment by flatmax** [comment]
I don't think I understand the sophistication of the 'next' callback yet 
... however ... I was thinking of something more along the following 
lines ....

Consider a Class with methods `{fn1, fn2, fn3 ... }`  ... can an expose 
method be automated so that it would do the following :

```js
test = new Class;
remote.expose(test);
```

is the equivalent of :
```js
remote.expose(
  {'test.fn1' : function(params, next){return next(false,test.fn1(params));},
  {'test.fn2' : function(params, next){return next(false,test.fn2(params));},
  {'test.fn3' : function(params, next){return next(false,test.fn3(params));},
)
```

thanks
Matt

**Comment by vphantom** [comment]
Yeah the issue here is that this would probably have to be a blocking (i.e. sync) implementation.  The binding to refer back to the instance `test` is feasible, but blocking is a run-time issue.

The motivation for callbacks is that it allows your method to do something slow (like I/O) asynchronously while your caller is also free to continue its execution at the same time.  The JS event loop takes care of dispatching execution to whatever waiting bits of code which are ready, as long as we give control back to it voluntarily (it's one of the downsides of JS).

FYI two advanced books which I found very useful in becoming fluent with JavaScript's intricacies and potential:
- [Effective JavaScript](http://effectivejs.com/) by David Herman
- [JavaScript, The Good Parts](http://shop.oreilly.com/product/9780596517748.do) By Douglas Crockford

All that being said, I'm keeping this ticket open to investigate if ES6 promises could be used to do something like what you're looking for.
